pyscf/dft2/libxc.py

```diff
@@ -0,0 +1,1231 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors: Qiming Sun <osirpt.sun@gmail.com>
+#          Susi Lehtola <susi.lehtola@gmail.com>
+# This file is adapted from `dft/libxc.py` of the PySCF core module.
+
+'''
+XC functional, the interface to libxc
+(http://www.tddft.org/programs/octopus/wiki/index.php/Libxc)
+'''
+
+import sys
+import warnings
+import math
+import numpy
+from functools import lru_cache
+from pyscf import lib
+from pyscf.dft.xc.utils import remove_dup, format_xc_code
+from pyscf.dft import xc_deriv
+from pyscf.dft.libxc import XC_CODES, XC, PROBLEMATIC_XC, XC_KEYS, XC_ALIAS, _NAME_WITH_DASH
+from pyscf import __config__
+try:
+    # API mode interface
+    from .libxc_cffi import ffi as _ffi, lib as _lib
+    _lib_eval_xc = _lib
+except ImportError:
+    # ABI mode interface
+    def cffi_abi_import():
+        global lib
+        import os
+        import cffi
+        from ._libxc_header import preprocess_header, load_file
+        pyscf_lib_path = os.path.dirname(lib.__file__)
+        forge_path = os.path.dirname(__file__)
+        header_file_path = os.path.join(pyscf_lib_path, 'deps', 'include', 'xc.h')
+        libxc_path = os.path.join(pyscf_lib_path, 'deps', 'lib', 'libxc.so')
+        eval_xc_path = os.path.join(forge_path, '..', 'lib', 'libxc_itrf2.so')
+        itrf_h_path = os.path.join(forge_path, '..', 'lib', 'dft', 'libxc_itrf2.h')
+
+        _ffi = cffi.FFI()
+        _ffi.cdef(preprocess_header(header_file_path) + load_file(itrf_h_path))
+        _lib = _ffi.dlopen(libxc_path)
+        _lib_eval_xc = _ffi.dlopen(eval_xc_path)
+        return _ffi, _lib, _lib_eval_xc
+    _ffi, _lib, _lib_eval_xc = cffi_abi_import()
+    del cffi_abi_import
+
+def libxc_version():
+    '''Returns the version of libxc'''
+    return _ffi.string(_lib.xc_version_string()).decode("UTF-8")
+
+def libxc_reference():
+    '''Returns the reference to libxc'''
+    return _ffi.string(_lib.xc_reference()).decode("UTF-8")
+
+def libxc_reference_doi():
+    '''Returns the reference to libxc'''
+    return _ffi.string(_lib.xc_reference_doi()).decode("UTF-8")
+
+__version__ = libxc_version()
+__reference__ = libxc_reference()
+__reference_doi__ = libxc_reference_doi()
+
+_XC_FUNC_TYPE_PTR_TYPE = _ffi.typeof("xc_func_type*")
+_XC_FUNC_TYPE_ARRAY_TYPE = _ffi.typeof("xc_func_type[]")
+_DOUBLE_PTR_TYPE = _ffi.typeof("double*")
+_DOUBLE_ARRAY_TYPE = _ffi.typeof("double[]")
+_DOUBLE_ARRAY_2_TYPE = _ffi.typeof("double[2]")
+_DOUBLE_ARRAY_3_TYPE = _ffi.typeof("double[3]")
+
+def xc_func_init(xc_code, nspin=_lib.XC_UNPOLARIZED):
+    func = _ffi.new(_XC_FUNC_TYPE_PTR_TYPE)
+    if _lib.xc_func_init(func, xc_code, nspin) != 0:
+        raise ValueError(f"Functional '{xc_code}' not found")
+    #  xc_func_end() will automatically be called when `func` is destructed
+    return _ffi.gc(func, _lib.xc_func_end)
```

Comment:
In my implementation, I used void* to pass the address of structs to LibXC. This is because LibXC already has many APIs to get and set functional properties. Therefore, I think exposing the members of the structs in Python is not really necessary. Only the address of the struct is important, and that is what void* does. If users really need access to the struct members that are not exposed by the API, they may either write a C library or explicitly declare the struct members in ctypes.